File: streams/examples/05-per-entity-hourly-profile.js
/**
 * Example 05 — Per-entity hourly profile build.
 *
 * Demonstrates the "per-entity stateful pipeline" shape: source events are
 * partitioned by customerId, each customer's events arrive FIFO ordered,
 * we accumulate an hourly profile per customer, and on hour boundary we
 * write the profile to a database via .foreach (at-least-once).
 *
 * If you need exactly-once external effects, prefer .to(queue) and have a
 * separate worker drain the sink with a transactional ack into your DB.
 *
 * Run:
 *   QUEEN_URL=http://localhost:6632 node examples/05-per-entity-hourly-profile.js
 */

import { Queen } from 'queen-mq'
import { Stream } from '@queenmq/streams'

const url = process.env.QUEEN_URL || 'http://localhost:6632'
const q = new Queen(url)

// Stand-in for your business DB write. In a real app this would be a
// pg.Client, a knex connection, or an HTTP call.
async function writeProfile(profile) {
  console.log('[example05] writing profile for', profile.customerId, '=>', JSON.stringify(profile))
  // await myPg.query('INSERT INTO customer_profiles ...', [...])
}

function mergeProfile(acc, m) {
  acc.customerId = m.data.customerId
  acc.events = (acc.events || 0) + 1
  acc.lastSeen = m.createdAt || new Date().toISOString()
  if (m.data.amount) {
    acc.totalAmount = (acc.totalAmount || 0) + m.data.amount
  }
  if (Array.isArray(m.data.tags)) {
    acc.tags = Array.from(new Set([...(acc.tags || []), ...m.data.tags]))
  }
  return acc
}

const handle = await Stream
  .from(q.queue('account.events'))
  // Default key = source partition_id (== customerId in this layout).
  .windowTumbling({ seconds: 3600 })
  .reduce(mergeProfile, { customerId: null, events: 0, totalAmount: 0, tags: [] })
  .foreach(profile => writeProfile(profile))
  .run({
    queryId: 'examples.customer.hourly_profile',
    url,
    batchSize: 100,
    maxPartitions: 4
  })

await new Promise(r => setTimeout(r, 60 * 60_000))
console.log('Stats:', handle.metrics())
await handle.stop()
await q.close()

File: clients/client-py/examples/05_per_entity_hourly_profile.py
"""
Example 05 — Per-entity hourly profile build. 1:1 port of 05-per-entity-hourly-profile.js.

Run:
    QUEEN_URL=http://localhost:6632 python examples/05_per_entity_hourly_profile.py
"""

from __future__ import annotations

import asyncio
import json
import os
from datetime import datetime, timezone

from queen import Queen, Stream


URL = os.environ.get("QUEEN_URL", "http://localhost:6632")


async def write_profile(profile: dict) -> None:
    print(f"[example05] writing profile for {profile.get('customerId')} => {json.dumps(profile)}")


def merge_profile(acc, m):
    data = m.get("data", {}) if isinstance(m, dict) else {}
    acc["customerId"] = data.get("customerId")
    acc["events"] = acc.get("events", 0) + 1
    acc["lastSeen"] = m.get("createdAt") or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    if data.get("amount"):
        acc["totalAmount"] = acc.get("totalAmount", 0) + data["amount"]
    if isinstance(data.get("tags"), list):
        acc["tags"] = list(set([*(acc.get("tags") or []), *data["tags"]]))
    return acc


async def main():
    q = Queen(url=URL)
    try:
        handle = await (
            Stream.from_(q.queue("account.events"))
            .window_tumbling(seconds=3600)
            .reduce(merge_profile, {"customerId": None, "events": 0, "totalAmount": 0, "tags": []})
            .foreach(lambda profile: write_profile(profile))
            .run(query_id="examples.customer.hourly_profile", url=URL, batch_size=100, max_partitions=4)
        )
        await asyncio.sleep(60 * 60)
        print("Stats:", handle.metrics())
        await handle.stop()
    finally:
        await q.close()


if __name__ == "__main__":
    asyncio.run(main())
